### PR TITLE
[ML/Train] Add `TensorFlowTrainer`

### PR DESCRIPTION
**Pull request by Yard1**
## Why are these changes needed?

Implements `TensorflowTrainer` & `TensorflowPredictor` for Ray AIR. Includes necessary modifications to `Dataset.to_tf`

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by clarkzinzow — APPROVED** [review]
Datasets changes look good, approving as a code owner for those changes

File: python/ray/train/examples/tensorflow_linear_dataset_example.py [hunk 1]
```diff
@@ -55,7 +55,7 @@ def train_func(config):
         # Model building/compiling need to be within `strategy.scope()`.
         multi_worker_model = build_and_compile_model(config)
 
-    dataset_pipeline = train.get_dataset_shard()
+    dataset_pipeline = train.get_dataset_shard(config.get("dataset_name", None))
```
Comment (amogkam) [inline python/ray/train/examples/tensorflow_linear_dataset_example.py]: Hmm what's this change for?

Comment (amogkam) [reply python/ray/train/examples/tensorflow_linear_dataset_example.py]: Oh I see this is for the test.

Comment (amogkam) [reply python/ray/train/examples/tensorflow_linear_dataset_example.py]: Can we use the function from `v2/examples` for the test instead of the one here?

**Review by amogkam — APPROVED** [review]
LGTM! Just one minor comment.

File: python/ray/train/examples/tensorflow_linear_dataset_example.py (hunk 1 above)
Comment (Yard1) [reply python/ray/train/examples/tensorflow_linear_dataset_example.py]: yeah will do

File: python/ray/train/v2/tests/test_trainer.py [hunk 2]
```diff
@@ -305,6 +313,66 @@ def train_func():
...
+            assert result[-1]["loss"] < result[0]["loss"]
+
+        num_workers = num_workers
+        epochs = 3
+        scaling_config = DataParallelScalingConfig(num_workers=num_workers)
+        config = {
+            "lr": 1e-3,
+            "batch_size": 32,
+            "epochs": epochs,
+            "dataset_name": TRAIN_DATASET_KEY,
```
Comment (amogkam) [inline python/ray/train/v2/tests/test_trainer.py]: Do we still need this now that the example is directly calling `train.get_train_dataset_shard()`?

Comment (Yard1) [reply python/ray/train/v2/tests/test_trainer.py]: nope, good catch

File: python/ray/data/dataset.py [hunk 3]
```diff
@@ -2188,8 +2188,8 @@ def get_feature_tensors(
     def to_tf(
```
Comment (ericl) [inline python/ray/data/dataset.py:2188]: Can you split this into a separate PR against datasets?

File: python/ray/ml/predictors/tensorflow.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
+import pandas as pd
```
Comment (ericl) [inline python/ray/ml/predictors/tensorflow.py:1]: Could we put integration-specific functionality in `ray.ml.integrations.*` or `ray.train.integrations.*`? I think cleanly separating integrations from core interfaces is important for readability.

File: python/ray/data/dataset.py (hunk 3 above)
Comment (Yard1) [reply python/ray/data/dataset.py:2188]: https://github.com/ray-project/ray/pull/22916
